### PR TITLE
test/alternator: clean up write isolation default and add more tests for the different modes

### LINKED ISSUE #24442: test/alternator: improve write isolation default and testing
From early on (commit c3da9f2bd4a767b53c58e09dbfc2642fa82a581e), the Alternator functional test suite (test/alternator) ran by default with write isolation mode `always_use_lwt`. The original thinking was that we need to exercise this more difficult mode and it's the most important mode. This mode was originally chosen in test/alternator/run.

However, starting with 76a766cab0e608167b813840d56014d036a1d066 (already two years ago, by @xtrey), test.py no longer runs test/alternator/run. Instead, it runs Scylla itself, and the options for running Scylla appear in test/alternator/suite.yaml, and for an unexplained reason the write isolation mode `only_rmw_uses_lwt` was chosen. So for two years, this is the mode used by all of our Jenkins tests.

We obviously didn't intend to change the default tested mode, but now that we're in this mess already, which mode **should** we use in all the tests? `always_use_lwt` or `only_rmw_uses_lwt`? By the way, when we started, we assumed that the former is the best and safest option and everyone will use it, but I recently learned that Scylla Cloud actually defaults to the latter, which is faster. 

Whatever mode we choose should be used in both test.py and test/alternator/run...

Additionally, I just realized we don't really have in-tree functional tests for all the modes. We have in test_tag.py some tests that it's possible to set these modes (without testing anything beyond setting the mode), and we do have a few functional tests for this issue in dtest's alternator_tests.py, but I would like to see in-tree in test/alternator also some tests that these write isolation modes are functionally correct - not regarding isolation - but other things - like `forbid_rmw` not being able to do any read-before write (condition expression, update expression requiring the item, returning old values, etc.) but can do updatexpression that doesn't require the item. Also verify that all other modes can correctly do (without testing isolation) all of the above types of writes (conditions, updates, return old, etc.).

### PR DESCRIPTION
**Pull request by nyh**
In #24442 it was noticed that accidentally, for a year now, test.py and CI were running the Alternator functional tests (test/alternator) using one write isolation mode (`only_rmw_uses_lwt`) while the manual test/alternator/run used a different write isolation mode (`always_use_lwt`). There is no good reason for this discrepancy, so in the second patch of this 2-patch series we change test/alternator/run to use the write isolation mode that we've had in CI for the last year.

But then, discussion on #24442 started: Instead of picking one mode or the other, don't we need test both modes? In fact, all four modes? 

The honest answer is that running **all tests** with **all combinations of options** is not practical - we'll find ourselves with an exponentially growing number of tests. What we really need to do is to run most tests that have nothing to do with write isolation modes on just one arbitrary write isolation mode like we're doing today. For example, numerous tests for the finer details of the ConditionExpression syntax will run on one mode. But then, have a separate test that verifies that one representative example of ConditionExpression (for example) works correctly on all four write isolation modes - rejected in forbid_rmw mode, allowed and behaves as expected on the other three. We had **some** tests like that in our test suite already, but the first patch in this series adds many more, making the test much more exhaustive and making it easier to review that we're really testing all four write isolation modes in every scenario that matters.

Fixes #24442 

No need to backport this patch - it's just adding more tests and changing developer-only test behavior.

### CODE REVIEW

File: test/alternator/run [hunk 1]
```diff
@@ -65,7 +65,7 @@ def run_alternator_cmd(pid, dir):
     cmd += [
         '--alternator-address', ip,
         '--alternator-enforce-authorization', '1',
-        '--alternator-write-isolation', 'always_use_lwt',
```
Comment (ScyllaPiotr) [inline test/alternator/run:68]: We ultimately never finished the discussion in the issue thread. With this change, you decide to move **all** the tests to default from fully consistent to eventually consistent (right?) and therefore susceptible to data inconsistencies.

File: test/alternator/test_write_isolation.py [hunk 2]
```diff
@@ -0,0 +1,430 @@
+# Copyright 2025-present ScyllaDB
+#
+# SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+
+# Tests for the the operations allowed under the four different write isolation
```
Comment (ScyllaPiotr) [inline test/alternator/test_write_isolation.py]: repetition: the the

File: test/alternator/test_write_isolation.py [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+# write isolation mode:
+rmw_forbidden='ValidationException.*write isolation policy'
+
+# The "test_table_s_*" fixtures are the same as test_table_s, except with the
+# given write isolation mode.
+# Because write isolation modes are a Scylla-only feature, these fixtures
+# are marked scylla_only, so all tests that use one of them are skipped
+# when not running against Scylla.
+@pytest.fixture(scope='module')
+def test_table_s_forbid(dynamodb, scylla_only):
```
Comment (ScyllaPiotr) [inline test/alternator/test_write_isolation.py]: I am generally a fan of elaborate names. I propose to use the following suffixes for the fuxtures:
- `forbid_rmw`
- `always_lwt`
- `only_rmw_lwt`
- `unsafe_rmw`

I find the current names ambiguous.

File: test/alternator/test_write_isolation.py [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+
+#############################################################################
+# "ConditionExpression" tests.
+# ConditionExpression applies to PutItem, DeleteItem and UpdateItem.
+# Without ConditionExpression, these operations work on all modes. With
+# ConditionExpression, they work on the three permit modes, and fail on
+# the forbid mode.
+
+# PutItem & ConditionExpression:
+def test_isolation_putitem_conditionexpression(test_table_s_forbid, test_table_s_permit):
```
Comment (ScyllaPiotr) [inline test/alternator/test_write_isolation.py]: This name I like :+1::smile:

File: test/alternator/test_write_isolation.py [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+            ConditionExpression='a = :oldval',
+            ExpressionAttributeValues={':oldval': 1})
+        assert table.get_item(Key={'p': p}, ConsistentRead=True)['Item'] == {'p': p, 'a': 2}
+    with pytest.raises(ClientError, match=rmw_forbidden):
+        test_table_s_forbid.put_item(Item={'p': p, 'a': 2},
+            ConditionExpression='a = :oldval',
+            ExpressionAttributeValues={':oldval': 1})
+
+# DeleteItem & ConditionExpression:
+def test_isolation_deleteitem_conditionexpression(test_table_s_forbid, test_table_s_permit):
```
Comment (ScyllaPiotr) [inline test/alternator/test_write_isolation.py]: Perhaps it would be beneficial (execution time-wise) to merge this test with the putitem one. There are many similar steps in the two tests.

File: test/alternator/test_write_isolation.py [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+    # Without ConditionExpression, DeleteItem works correctly on all modes
+    p = random_string()
+    for table in test_table_s_permit + [test_table_s_forbid]:
+        table.put_item(Item={'p': p, 'a': 1})
+        table.delete_item(Key={'p': p})
```
Comment (ScyllaPiotr) [inline test/alternator/test_write_isolation.py:147]: If you agreed to merge this with the putitem test, skip this comment,
otherwise,
I recall that in other places you usually make sure the data is there before attempting to remove it. I understand that currently there is another test that given similar code makes sure the data is put there, but one could also claim that this test does not do good in making sure the data is actually being deleted (since we never checked if it was written successfully).
Another, distinct line of thinking is along your earlier comment: "_We only test here which operations are allowed or not allowed in each write isolation mode_". Along these lines, you should not need to check if data has been actually deleted. The final lines of this test demonstrate this approach (although for a different reason - an exception is expected).
:bulb: No, actually you must have thought about reusing the same table in different tests.

File: test/alternator/test_write_isolation.py [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+        assert 'Item' not in table.get_item(Key={'p': p}, ConsistentRead=True)
+    # With ConditionExpression, DeleteItem works correctly on permit modes,
+    # fails on forbid mode.
+    for table in test_table_s_permit:
+        table.put_item(Item={'p': p, 'a': 1})
+        table.delete_item(Key={'p': p},
```
Comment (ScyllaPiotr) [inline test/alternator/test_write_isolation.py:154]: ditto

File: test/alternator/test_write_isolation.py (hunk 4 above)
Comment (ScyllaPiotr) [reply test/alternator/test_write_isolation.py]: After this test, the data stays in the tables? Even more incentive to merge this test with the deleteitem one, that I write about further down.

File: test/alternator/test_write_isolation.py [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+            ConditionExpression='a = :oldval',
+            ExpressionAttributeValues={':oldval': 1})
+        assert 'Item' not in table.get_item(Key={'p': p}, ConsistentRead=True)
+    with pytest.raises(ClientError, match=rmw_forbidden):
+        test_table_s_forbid.delete_item(Key={'p': p},
```
Comment (ScyllaPiotr) [inline test/alternator/test_write_isolation.py]: You test legality of conditional deletions with different data in the tables inside `test_table_s_permit` vs `test_table_s_forbid`. It's not a problem, but an actual difference in the test conditions.

File: test/alternator/test_write_isolation.py [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+            ConditionExpression='a = :oldval',
+            ExpressionAttributeValues={':oldval': 1})
+
+# UpdateItem & ConditionExpression:
+# Note that an UpdateItem always needs some sort of update expression, so
+# we deliberately test with an update expression that doesn't itself need
+# a read-before-write (we'll check more elaborate UpdateExpression below).
+def test_isolation_updateitem_conditionexpression(test_table_s_forbid, test_table_s_permit):
```
Comment (ScyllaPiotr) [inline test/alternator/test_write_isolation.py]: After this test, and the others further down, the data stays in the tables?

File: test/alternator/test_write_isolation.py [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+        assert 'Item' not in table.get_item(Key={'p': p}, ConsistentRead=True)
+    # With ReturnValues=ALL_OLD, DeleteItems works correctly (also returning
+    # the old item) on permit modes, and fails on forbid mode.
+    for table in test_table_s_permit:
+        table.put_item(Item={'p': p, 'a': 1})
+        ret = table.delete_item(Key={'p': p}, ReturnValues='ALL_OLD')
+        assert ret['Attributes'] == {'p': p, 'a': 1}
+        assert 'Item' not in table.get_item(Key={'p': p}, ConsistentRead=True)
+    with pytest.raises(ClientError, match=rmw_forbidden):
+        test_table_s_forbid.delete_item(Key={'p': p}, ReturnValues='ALL_OLD')
```
Comment (ScyllaPiotr) [inline test/alternator/test_write_isolation.py]: You test with different data in the tables inside `test_table_s_permit` vs the `test_table_s_forbid`. It's not a problem, but an actual difference in the test conditions.

File: test/alternator/test_write_isolation.py [hunk 11]
```diff
@@ -0,0 +1,430 @@
...
+# so work on all write isolation modes except forbid_rmw
+def test_isolation_attributeupdates_rmw(test_table_s_forbid, test_table_s_permit):
+    p = random_string()
+    for table in test_table_s_permit:
+        table.put_item(Item={'p': p, 'a': 1})
+        table.update_item(Key={'p': p},
+            AttributeUpdates={'a': {'Value': 1, 'Action': 'ADD'}})
+        assert table.get_item(Key={'p': p}, ConsistentRead=True)['Item'] == {'p': p, 'a': 2}
+    with pytest.raises(ClientError, match=rmw_forbidden):
+        test_table_s_forbid.update_item(Key={'p': p},
```
Comment (ScyllaPiotr) [inline test/alternator/test_write_isolation.py]: You test with different data in the tables inside `test_table_s_permit` vs the `test_table_s_forbid`. It's not a problem, but an actual difference in the test conditions.

File: test/alternator/test_write_isolation.py [hunk 12]
```diff
@@ -0,0 +1,430 @@
...
+    # With Expected, DeleteItem works correctly on permit modes,
+    # fails on forbid mode.
+    p = random_string()
+    for table in test_table_s_permit:
+        table.put_item(Item={'p': p, 'a': 1})
+        table.delete_item(Key={'p': p},
+            Expected={'a': {'ComparisonOperator': 'EQ', 'AttributeValueList': [1]}})
+        assert 'Item' not in table.get_item(Key={'p': p}, ConsistentRead=True)
+    with pytest.raises(ClientError, match=rmw_forbidden):
+        test_table_s_forbid.delete_item(Key={'p': p},
```
Comment (ScyllaPiotr) [inline test/alternator/test_write_isolation.py]: You test `delete_item` with different data in the tables inside `test_table_s_permit` vs the `test_table_s_forbid`. It's not a problem, but an actual difference in the test conditions.

File: test/alternator/test_write_isolation.py [hunk 13]
```diff
@@ -0,0 +1,430 @@
...
+    # fails on forbid mode.
+    p = random_string()
+    for table in test_table_s_permit:
+        table.put_item(Item={'p': p, 'a': 1})
+        table.update_item(Key={'p': p},
+            AttributeUpdates={'a': {'Value': 2, 'Action': 'PUT'}},
+            Expected={'a': {'ComparisonOperator': 'EQ', 'AttributeValueList': [1]}})
+        assert table.get_item(Key={'p': p}, ConsistentRead=True)['Item'] == {'p': p, 'a': 2}
+    with pytest.raises(ClientError, match=rmw_forbidden):
+        test_table_s_forbid.update_item(Key={'p': p},
```
Comment (ScyllaPiotr) [inline test/alternator/test_write_isolation.py]: You test `update_item` with different data in the tables inside `test_table_s_permit` vs the `test_table_s_forbid`. It's not a problem, but an actual difference in the test conditions.

File: test/alternator/test_write_isolation.py [hunk 14]
```diff
@@ -0,0 +1,430 @@
...
+    # With Expected, PutItem works correctly on permit modes,
+    # fails on forbid mode.
+    p = random_string()
+    for table in test_table_s_permit:
+        table.put_item(Item={'p': p, 'a': 1})
+        table.put_item(Item={'p': p, 'a': 2},
+            Expected={'a': {'ComparisonOperator': 'EQ', 'AttributeValueList': [1]}})
+        assert table.get_item(Key={'p': p}, ConsistentRead=True)['Item'] == {'p': p, 'a': 2}
+    with pytest.raises(ClientError, match=rmw_forbidden):
+        test_table_s_forbid.put_item(Item={'p': p, 'a': 2},
```
Comment (ScyllaPiotr) [inline test/alternator/test_write_isolation.py]: You test `put_item` with different data in the tables inside `test_table_s_permit` vs the `test_table_s_forbid`. It's not a problem, but an actual difference in the test conditions.

**Review by ScyllaPiotr — CHANGES_REQUESTED** [review]
Wow, very nice suite of tests!
I only raised a couple of concerns, I wonder if you will share them.

File: test/alternator/run (hunk 1 above)
Comment (nyh) [reply test/alternator/run:68]: It's not exactly "fully consistent vs eventually consistent". I tried to explain the differences between these modes in the documentation, but to make a long story short "always_use_lwt" uses LWT for every write and guarantees that any concurrent writes are isolated - even if you mix RMW and pure writes (e.g., consider i=3 and doing "i=7" and "i++" in parallel. With correct isolation you end up with either i=7 or i=8. With incorrect isolation you can end up with i=4.). The mode "only_rmw_use_lwt" only isolates RMW, and assumes you know you don't mix pure writes and RMW to the same item - which is an assumption that DynamoDB doesn't make, but some users know is true.

Anyway, for the **tests** it doesn't really matter what mode to use. As I noted as an example, test_condition_expression.py has almost 2,000 lines of code, 53 test functions, to test a large number of syntax features and corner cases of condition expression. None of these tests care how we do write isolation. We just need one test (that I cleaned up in this patch) to check how condition expressions - any condition expressions - work with the different write isolation modes, and all the 53 other tests - they can run on any mode we choose. We don't need to choose the mode after very deliberate thinking.

File: test/alternator/test_write_isolation.py (hunk 2 above)
Comment (nyh) [reply test/alternator/test_write_isolation.py]: Thanks. Fixed.

File: test/alternator/test_write_isolation.py (hunk 3 above)
Comment (nyh) [reply test/alternator/test_write_isolation.py]: Perhaps. I originally decided on these shorter names because:
1. These names are used dozens of times throughout the test code, e.g., ```test_table_s_forbid.put_item(Item={'p': p, 'a': 2},```, and the longer they are the harder it is to read those lines - I think.
2. Alternator already supports any subset of these names, so instead of "forbid_rmw" you can use "forbid" or even "f" and this is even documented.

File: test/alternator/test_write_isolation.py (hunk 4 above)
Comment (nyh) [reply test/alternator/test_write_isolation.py]: Yes, all throughout the Alternator test suite, tests share the same tables, and add data to a unique key, and often this data just stays in the table which doesn't matter at all - and is probably even faster than bothering to delete the items with additional comments. Instead, the entire is delete in one fell swoop when the test ends (for the tables like test_table_s shared by all tests) or in the case of the tables used only in this test file, in a "module"-scoped fixture, the table is deleted at the end of the file.

So there is no penalty for leaving a live item.
You're right that there's a small penalty when two tests need to create the same item (and each test creates the item in a different key) but this penality is absolutely tiny - less than a millisecond, and not worth optimizing from.

File: test/alternator/test_write_isolation.py (hunk 5 above)
Comment (nyh) [reply test/alternator/test_write_isolation.py]: As I noted above perhaps we can save a couple of milliseconds this way, but I don't think it's worth it.

File: test/alternator/test_write_isolation.py (hunk 6 above)
Comment (nyh) [reply test/alternator/test_write_isolation.py:147]: When I wrote "We only test here which operations are allowed or not allowed in each write isolation mode", by "allowed" I meant more like "honored" - i.e., allowed and actually performed. Obviously if the operation just didn't print an error but didn't do the actual operation - that would not be good enough. So for example I wanted to test that DeleteItem with ConditionExpression not only is allowed in "unsafe" mode (or whatever), it is actually performed when the conditionexpression is true - meaning that the read-before-write actually worked. I want to check that the read-before-write actually worked in all modes, for all operations, and that's the goal of this test.

You're right that if put_item() didn't even work, the delete_item will seem to "work" and delete the item that was never added. I guess you're right it's safer against test bugs to add an assert that the item exists between the put_item and delete_item. In this case I guess I thought checking that put_item worked is a bit too silly, but perhaps you're right it isn't - I can add those extra tests.

I'm not sure what the final 💡line means - yes, I do deliberately use the same tables (test_table_s_forbid and its three friends in this test) because table creation and deletion is the slowest part of each test so I want to avoid to do it separately in each test. Sharing of tables between tests is one of the reasons why the Alternator tests are so blazingly fast - may take less than 1/100th of a second, the minimum that pytest's "--durations=0" measures  :-)

File: test/alternator/test_write_isolation.py (hunk 8 above)
Comment (nyh) [reply test/alternator/test_write_isolation.py]: You're right: In the "permit" modes I wanted to check that the operation is not only permitted but actually performed - so I need the condition to succeed so that the operation is performed. So I did a putitem to make the condition succeed but only if the read-before-write is actually done correctly.

But for "forbid" mode the operation should just be forbidden (with the special RMW forbidden error) - it doesn't actually matter what the condition or the data is.

I didn't think this is a problem, if you think it's weird I can easily add a put_item in the "forbid" case as well.

File: test/alternator/test_write_isolation.py (hunk 9 above)
Comment (nyh) [reply test/alternator/test_write_isolation.py]: Yes, I tried to explain this above, but if it's still not clear I'll be happy to explain a bit about the "philosophy" of the alternator test suite in one of our weekly calls. Briefly, the idea is that it should be easy to write these tests and that they should be super quick (super quick means on average 50ms or less per test, several orders of magnitude faster than dtest). I believe that both features encourage developers like me to actually write these tests instead of avoiding them.

File: test/alternator/test_write_isolation.py (hunk 12 above)
Comment (nyh) [reply test/alternator/test_write_isolation.py]: Yes, I answered this above as well. If you feel this is a problem I can change both^H^H^H^Hall four places.

File: test/alternator/test_write_isolation.py (hunk 13 above)
Comment (nyh) [reply test/alternator/test_write_isolation.py]: Ditto.

File: test/alternator/test_write_isolation.py (hunk 14 above)
Comment (nyh) [reply test/alternator/test_write_isolation.py]: Ditto.

File: test/alternator/run (hunk 1 above)
Comment (ScyllaPiotr) [reply test/alternator/run:68]: > It's not exactly "fully consistent vs eventually consistent".

That's clear now, thanks for the good example.

> Anyway, for the tests it doesn't really matter what mode to use.

Well, the issue cover letter asks which isolation mode should we choose, so it must be relevant, at least - between the test suites (`test.py` vs `test/alternator/run`).

File: test/alternator/test_write_isolation.py (hunk 3 above)
Comment (ScyllaPiotr) [reply test/alternator/test_write_isolation.py]: Ad. 1 Well, new devs obviously have less context to base upon and for them longer (more verbose) names are easier to read (understand).

Ad 2. Well I'm not sure it was a good idea to allow these abbreviations, but it's pointless to exchange arguments about it now. However, the way one **uses** it now can be argued and a lone `forbid` is ambiguous, while `f` for me is blatantly misleading - for me it associates with _force_.

But it's obviously only an opinion.

File: test/alternator/test_write_isolation.py (hunk 4 above)
Comment (ScyllaPiotr) [reply test/alternator/test_write_isolation.py]: I had thought about an influence on correctness of the tests, assuming the tests are run (at least statistically) in a specific order.

File: test/alternator/test_write_isolation.py (hunk 6 above)
Comment (ScyllaPiotr) [reply test/alternator/test_write_isolation.py:147]: Great!

The 💡 was in line with what you wrote in the 1st paragraph - I assumed you want to check if the `put_item` is only "honored" and never verified actually doing its job. Therefore the `delete_item` was suspected by me only to purge the data for the subsequent test.

File: test/alternator/test_write_isolation.py (hunk 8 above)
Comment (ScyllaPiotr) [reply test/alternator/test_write_isolation.py]: It's not weird nor incorrect, but inconsistent. I would propose to make it consistent.

File: test/alternator/test_write_isolation.py (hunk 9 above)
Comment (ScyllaPiotr) [reply test/alternator/test_write_isolation.py]: No, the question was indirect, it was meant as a trigger to actually re-verify this. I understand the rationale, only not entirely the consequences for the whole set of affected tests.

But perhaps it could be beneficial to others to actually learn about this too on (one) Thursday :smile:

**Comment by nyh** [comment]
>     * [test_random_failures[stop_after_starting_group0_service-add_new_udt].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17716/testReport/junit/cluster.random_failures/test_random_failures/Tests___Unit_Tests___test_random_failures_stop_after_starting_group0_service_add_new_udt__debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures%5Bstop_after_starting_group0_service-add_new_udt%5D.debug.2)

Known flakiness - this is https://github.com/scylladb/scylladb/issues/24513 - a recent regression that Avi had to revert. I'll rebase to get the newer master with the revert.

File: test/alternator/test_write_isolation.py (hunk 3 above)
Comment (nyh) [reply test/alternator/test_write_isolation.py]: Ok, I'll change in this test to the long names (I won't change the API). I want this PR to go in :-)

Comment (nyh) [reply test/alternator/test_write_isolation.py]: While at it, I'll remove the word "test" from the table fixture. While it made semantic sense when I called the original table in these tests "test_table" (it's a test table, after all!), it's really confusing for seasoned pytest developers who are used to test_something being the name of a test, not a fixture. I don't want to overhaul the entire test suite to rename test_table, but here we create four new tables, so I can give them a better name like table_s_forbid_rmw.

Or even shorter - just table_forbid_rmw. It's not like I have in this file tables with a different schema.

File: test/alternator/test_write_isolation.py (hunk 4 above)
Comment (nyh) [reply test/alternator/test_write_isolation.py]: That's why Alternator tests are all written to assume they do not work on a brand new table, so should work on a unique key (currently "random", if you're super-super-super unlucky it might not be unique, see https://github.com/scylladb/scylladb/issues/9988) and not a fixed key like "x".
A minority of the tests that do need very unique or brand-new tables can create a new table in the test, but most don't. Depending on your disk setup (/tmp or current directory), this testing philosophy can save 0.01 to 0.1 seconds on each test, which adds up when you have a thousand tests.

File: test/alternator/test_write_isolation.py (hunk 6 above)
Comment (nyh) [reply test/alternator/test_write_isolation.py:147]: Added the extra assert that the putitem worked.

File: test/alternator/test_write_isolation.py (hunk 7 above)
Comment (nyh) [reply test/alternator/test_write_isolation.py:154]: Added the extra assert after the put_item.

File: test/alternator/test_write_isolation.py (hunk 9 above)
Comment (nyh) [reply test/alternator/test_write_isolation.py]: Note how each test starts with `p = random_string()` (and some tests have several of these, when they want to test several different things).
This means the test picks a random key, and writes and reads to it. Such a test couldn't care less if the table has 100 other keys - it never scans the entire table, it just access the one unique key "p" it chose.

Feel free to invite me to a call one day and I'll prepare a presentation on my (or Alternator's) functional testing philosophy.

Comment (nyh) [reply test/alternator/test_write_isolation.py]: By the way, my presentation on my view of testing (which isn't the entire company's, but it's not very far) that I presented two years ago: https://docs.google.com/presentation/d/1q0iqAG7PcChkFtqhbtvU_Fj0mSdXuOWqMJWIwP2Yzew/edit?usp=sharing

**Comment by nyh** [comment]
Pushed a new version implementing most of @ScyllaPiotr 's suggestions about renaming the test tables etc, but didn't act on a bunch of places where you told me the test for "forbid" and "permit" modes are not identical (I know, and I think it's fine. the intention was to **test** all four modes - it's not necessary that the tests will be 100% identical).

@ScyllaPiotr you left "request changes" so please re-review.

File: test/alternator/test_write_isolation.py (hunk 9 above)
Comment (ScyllaPiotr) [reply test/alternator/test_write_isolation.py]: Wow, thanks!

File: test/alternator/test_write_isolation.py (hunk 8 above)
Comment (ScyllaPiotr) [reply test/alternator/test_write_isolation.py]: I still don't like it, but maybe it's too subjective and at the same time insignificant to block the whole PR.

**Comment by nyh** [comment]
@scylladb/scylla-maint please consider merging: There is one reviewer approval and the CI passed.

**Comment by nyh** [comment]
> @scylladb/scylla-maint please consider merging: There is one reviewer approval and the CI passed.

ping
